### PR TITLE
CS: fully_qualified_strict_types - import_symbols

### PR DESCRIPTION
**Pull request by keradus**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | no 
| Deprecations? | no 
| Issues        | Fix CS 
| License       | MIT

suggested in https://github.com/symfony/symfony/pull/53233/files#r1437244439

### CODE REVIEW

**Review by xabbuh — CHANGES_REQUESTED** [review]
failures look related to me

**Comment by keradus** [comment]
this PR serves a purpose to demonstrate "how the codebase would look alike if we would incorporate another Fixer rule", as one of members suggested to get it enabled.

Before investing any time into failure of this PR, I would like to get confirmation we want to go this path and enable the rule. Diff is massive and I do not want to invest time into investigation, if final conclusion would be that "we do not want this rule to be enabled".

What's your take on the rule itself?

**Comment by xabbuh** [comment]
importing the classes looks indeed like what we want to have in the codebase

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 1]
```diff
@@ -12,6 +12,7 @@
 namespace Symfony\Bundle\FrameworkBundle\DependencyInjection;
 
 use Composer\InstalledVersions;
+use GuzzleHttp\UriTemplate\UriTemplate;
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]:
```suggestion
use GuzzleHttp\UriTemplate\UriTemplate as GuzzleUriTemplate;
```

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 2]
```diff
@@ -23,12 +24,14 @@
 use Psr\Container\ContainerInterface as PsrContainerInterface;
 use Psr\Http\Client\ClientInterface;
 use Psr\Log\LoggerAwareInterface;
+use Rize\UriTemplate;
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]:
```suggestion
use Rize\UriTemplate as RizeUriTemplate;
```

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 3]
```diff
@@ -2432,9 +2438,9 @@ private function registerHttpClientConfiguration(array $config, ContainerBuilder
             $this->registerRetryableHttpClient($retryOptions, 'http_client', $container);
         }
 
-        if (ContainerBuilder::willBeAvailable('guzzlehttp/uri-template', \GuzzleHttp\UriTemplate\UriTemplate::class, [])) {
+        if (ContainerBuilder::willBeAvailable('guzzlehttp/uri-template', UriTemplate::class, [])) {
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]:
```suggestion
        if (ContainerBuilder::willBeAvailable('guzzlehttp/uri-template', GuzzleUriTemplate::class, [])) {
```

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 4]
```diff
@@ -2432,9 +2438,9 @@ private function registerHttpClientConfiguration(array $config, ContainerBuilder
...
             $container->setAlias('http_client.uri_template_expander', 'http_client.uri_template_expander.guzzle');
-        } elseif (ContainerBuilder::willBeAvailable('rize/uri-template', \Rize\UriTemplate::class, [])) {
+        } elseif (ContainerBuilder::willBeAvailable('rize/uri-template', UriTemplate::class, [])) {
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]:
```suggestion
        } elseif (ContainerBuilder::willBeAvailable('rize/uri-template', RizeUriTemplate::class, [])) {
```

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 2 above)
Comment (derrabus) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I would consider this to be a bug in the fixer: The fixer should detect that a class with the same name has already been imported from a different namespace and either do nothing or generate an alias for the import. Currently, the fixer would generate obviously broken code.

**Comment by keradus** [comment]
I take the eagerness to use the rule.

I will covert PR to draft as it is not in merge-ready state.

**Comment by Wirone** [comment]
Related to PHP-CS-Fixer/PHP-CS-Fixer#7629 - this PR probably could prepare a Symfony codebase for that.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 2 above)
Comment (Wirone) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: @derrabus yes, I agree, obviously it's a bug in the importing logic. I overlooked it in my PR 🙁.

Comment (Wirone) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Actually there was a test case for names collision, but it involved existing import + new symbol, while the problem here was related to names collision between newly found symbols, that I did not take into consideration. I've prepared fix for that in PHP-CS-Fixer/PHP-CS-Fixer#7635 🙂.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 3 above)
Comment (keradus) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: (outdated)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 4 above)
Comment (keradus) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: (outdated)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 1 above)
Comment (keradus) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: (outdated)

File: src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php [hunk 5]
```diff
@@ -132,7 +133,7 @@ public function __construct(array $options, \Redis|Relay|\RedisCluster|null $red
                             if ($address = $sentinel->getMasterAddrByName($sentinelMaster)) {
                                 [$host, $port] = $address;
                             }
-                        } catch (\RedisException|\Relay\Exception $redisException) {
+                        } catch (\RedisException|RelayException $redisException) {
```
Comment (keradus) [inline src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:136]: looks like Psalm cannot handle this, should I bother by that and stop the PR or continue with it?
cc @xabbuh 

also, reported here: https://github.com/vimeo/psalm/issues/10840

**Comment by nicolas-grekas** [comment]
Please let me close this one.
We're fine with doing this case by case.
Usually importing is the best way, but in some cases, inlining the FQCN provides more clarity to the codebase.

**Pull request closed without merging**